File: EEntityCore.DB/EEntityCore.DB/Generic/RecordSet.vb

Namespace Genric

    ''' <summary>
    ''' Helps to Navigate a Record Set
    ''' </summary>
    ''' <remarks></remarks>
    Public NotInheritable Class RecordSet(Of T)

        REM Currently Delete will not be implemented
        REM Just Load Data
        REM Navigate to and From

        REM When a Data is Loaded, specify current index should be kept or move to first if there is




#Region "Constructors"


        Sub New(ByVal DBCursorList() As T)

            Me.DBCursorList = DBCursorList
            Me.moveFirst()

        End Sub



#End Region


#Region "Properties"


        Public ReadOnly Property DBCursor As T
            Get
                If Me.CurrentIndex >= 0 AndAlso Me.hasRecords AndAlso Me.CurrentIndex < Me.DBCursorList.Count Then
                    Return Me.DBCursorList(Me.CurrentIndex)
                End If
                Return Nothing
            End Get
        End Property

        Private CurrentIndex As Int32 = -1

        Private DBCursorList() As T = Nothing


        Public ReadOnly Property hasRecords As Boolean
            Get
                Return Me.DBCursorList IsNot Nothing AndAlso Me.DBCursorList.Count > 0
            End Get
        End Property


        Public ReadOnly Property hasNext As Boolean
            Get
                Return Me.hasRecords AndAlso Me.CurrentIndex < Me.DBCursorList.Count - 1
            End Get
        End Property

        Public ReadOnly Property hasPrevious As Boolean
            Get
                Return Me.hasRecords AndAlso Me.CurrentIndex > 0
            End Get
        End Property

        ''' <summary>
        ''' Indicate it has reach eof of record
        ''' </summary>
        ''' <value></value>
        ''' <returns></returns>
        ''' <remarks></remarks>
        Public ReadOnly Property EOF As Boolean
            Get
                If Me.DBCursorList Is Nothing Then Return True
                Return Me.CurrentIndex + 1 > Me.DBCursorList.Count
            End Get
        End Property


        ''' <summary>
        ''' Indicate Position is at -1 index
        ''' </summary>
        ''' <value></value>
        ''' <returns></returns>
        ''' <remarks></remarks>
        Public ReadOnly Property BOF As Boolean
            Get
                Return Me.CurrentIndex = -1
            End Get
        End Property

        ''' <summary>
        ''' Displays the Absolute index of this current record. Like 1 of 4
        ''' </summary>
        ''' <value></value>
        ''' <returns></returns>
        ''' <remarks></remarks>
        Public ReadOnly Property AbsoluteIndex As Int32
            Get
                REM since there can be minimum of the Lowest digit
                REM Else you have 0 of 0
                If Not Me.EOF Then
                    Return Me.CurrentIndex + 1

                Else

                    Return 0

                End If

            End Get
        End Property


        Public ReadOnly Property TotalRecords As Int32
            Get
                If Me.hasRecords Then Return Me.DBCursorList.Count
                Return 0
            End Get
        End Property

#End Region



#Region "Methods"

        Private Sub NullPointer()
            Me.CurrentIndex = -1
        End Sub

        Public Function moveFirst() As T
            Me.NullPointer()
            If Me.hasRecords Then Me.CurrentIndex = 0
            Return Me.DBCursor
        End Function

        Public Function movePrevious() As T
            Dim __CurrentIndex As Int32 = Me.CurrentIndex
            If Me.hasRecords And __CurrentIndex > 0 Then
                If Me.EOF Then
                    Me.CurrentIndex = Me.DBCursorList.Count - 1
                Else
                    Me.CurrentIndex -= 1
                End If
            Else
                Me.NullPointer() REM No Record at all
            End If
            Return Me.DBCursor
        End Function



        Public Function moveNext() As T
            REM If I do it like this .. It will return the Cursor to the BOF when it reaches EOF
            ''If Me.hasNext Then
            ''    Me.CurrentIndex += 1
            ''Else
            ''    Me.NullPointer()
            ''End If
            Me.CurrentIndex += 1
            Return Me.DBCursor
        End Function

        Public Function moveLast() As T
            Me.NullPointer()
            If Me.hasRecords Then Me.CurrentIndex = Me.DBCursorList.Count - 1
            Return Me.DBCursor
        End Function

        Private Function gotoRecord(ByVal index As Int32) As T
            Me.NullPointer()
            If Me.hasRecords AndAlso index < Me.DBCursorList.Count Then Me.CurrentIndex = index
            Return Me.DBCursor

        End Function


        Public Sub reloadData(ByVal DBCursorList() As T,
                              Optional ByVal maintainCurrentIndex As Boolean = False
                                    )

            Dim __currentIndex As Int32 = Me.CurrentIndex
            Me.DBCursorList = DBCursorList
            If maintainCurrentIndex Then Me.gotoRecord(__currentIndex)

        End Sub

#End Region




    End Class

End Namespace

File: EEntityCore.DB/EEntityCore.DB/Generic/RecordSet.cs
using System.Linq;

namespace EEntityCore.DB.Genric
{

    /// <summary>
    /// Helps to Navigate a Record Set
    /// </summary>
    /// <remarks></remarks>
    public sealed class RecordSet<T>
    {

        // REM Currently Delete will not be implemented
        // REM Just Load Data
        // REM Navigate to and From

        // REM When a Data is Loaded, specify current index should be kept or move to first if there is




        #region Constructors


        public RecordSet(T[] DBCursorList)
        {
            this.DBCursorList = DBCursorList;
            moveFirst();
        }



        #endregion


        #region Properties


        public T DBCursor
        {
            get
            {
                if (CurrentIndex >= 0 && hasRecords && CurrentIndex < DBCursorList.Count())
                {
                    return DBCursorList[CurrentIndex];
                }

                return default;
            }
        }

        private int CurrentIndex = -1;
        private T[] DBCursorList = null;

        public bool hasRecords
        {
            get
            {
                return DBCursorList is object && DBCursorList.Count() > 0;
            }
        }

        public bool hasNext
        {
            get
            {
                return hasRecords && CurrentIndex < DBCursorList.Count() - 1;
            }
        }

        public bool hasPrevious
        {
            get
            {
                return hasRecords && CurrentIndex > 0;
            }
        }

        /// <summary>
        /// Indicate it has reach eof of record
        /// </summary>
        /// <value></value>
        /// <returns></returns>
        /// <remarks></remarks>
        public bool EOF
        {
            get
            {
                if (DBCursorList is null)
                    return true;
                return CurrentIndex + 1 > DBCursorList.Count();
            }
        }


        /// <summary>
        /// Indicate Position is at -1 index
        /// </summary>
        /// <value></value>
        /// <returns></returns>
        /// <remarks></remarks>
        public bool BOF
        {
            get
            {
                return CurrentIndex == -1;
            }
        }

        /// <summary>
        /// Displays the Absolute index of this current record. Like 1 of 4
        /// </summary>
        /// <value></value>
        /// <returns></returns>
        /// <remarks></remarks>
        public int AbsoluteIndex
        {
            get
            {
                // REM since there can be minimum of the Lowest digit
                // REM Else you have 0 of 0
                if (!EOF)
                {
                    return CurrentIndex + 1;
                }
                else
                {
                    return 0;
                }
            }
        }

        public int TotalRecords
        {
            get
            {
                if (hasRecords)
                    return DBCursorList.Count();
                return 0;
            }
        }

        #endregion



        #region Methods

        private void NullPointer()
        {
            CurrentIndex = -1;
        }

        public T moveFirst()
        {
            NullPointer();
            if (hasRecords)
                CurrentIndex = 0;
            return DBCursor;
        }

        public T movePrevious()
        {
            int __CurrentIndex = CurrentIndex;
            if (hasRecords & __CurrentIndex > 0)
            {
                if (EOF)
                {
                    CurrentIndex = DBCursorList.Count() - 1;
                }
                else
                {
                    CurrentIndex -= 1;
                }
            }
            else
            {
                NullPointer();
            } // REM No Record at all

            return DBCursor;
        }

        public T moveNext()
        {
            // REM If I do it like this .. It will return the Cursor to the BOF when it reaches EOF
            // 'If Me.hasNext Then
            // '    Me.CurrentIndex += 1
            // 'Else
            // '    Me.NullPointer()
            // 'End If
            CurrentIndex += 1;
            return DBCursor;
        }

        public T moveLast()
        {
            NullPointer();
            if (hasRecords)
                CurrentIndex = DBCursorList.Count() - 1;
            return DBCursor;
        }

        private T gotoRecord(int index)
        {
            NullPointer();
            if (hasRecords && index < DBCursorList.Count())
                CurrentIndex = index;
            return DBCursor;
        }

        public void reloadData(T[] DBCursorList, bool maintainCurrentIndex = false)
        {
            int __currentIndex = CurrentIndex;
            this.DBCursorList = DBCursorList;
            if (maintainCurrentIndex)
                gotoRecord(__currentIndex);
        }

        #endregion




    }
}